Blish HUD/GameServices/ArcDps/SocketListener.cs

```diff
@@ -0,0 +1,211 @@
+
+namespace Blish_HUD.ArcDps
+{
+    public sealed class SocketListener
+    {
+        public delegate void Message(MessageData data);
+
+        private const int MESSAGE_HEADER_SIZE = 8;
+
+        private static readonly Mutex Mutex = new Mutex();
+        private readonly Semaphore _acceptedClientsSemaphore;
+        private readonly int _bufferSize;
+        private readonly int _maxConnectionCount;
+
+        private readonly SocketAsyncEventArgsPool _socketAsyncReceiveEventArgsPool;
+        private Socket _listenSocket;
+
+        public SocketListener(int maxConnectionCount, int bufferSize)
+        {
+            _maxConnectionCount = maxConnectionCount;
+            _bufferSize = bufferSize;
+            _socketAsyncReceiveEventArgsPool = new SocketAsyncEventArgsPool();
+            _acceptedClientsSemaphore = new Semaphore(maxConnectionCount, maxConnectionCount);
+
+            for (var i = 0; i < maxConnectionCount; i++)
+            {
+                var socketAsyncEventArgs = new SocketAsyncEventArgs();
+                socketAsyncEventArgs.Completed += OnIoCompleted;
+                socketAsyncEventArgs.SetBuffer(new byte[bufferSize], 0, bufferSize);
+                _socketAsyncReceiveEventArgsPool.Push(socketAsyncEventArgs);
+            }
+        }
+
+        public event Message ReceivedMessage;
+
+        public void Start(IPEndPoint localEndPoint)
+        {
+            _listenSocket = new Socket(localEndPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp)
+            {
+                ReceiveBufferSize = _bufferSize
+            };
+            _listenSocket.Bind(localEndPoint);
+            _listenSocket.Listen(_maxConnectionCount);
+            StartAccept(null);
+            Mutex.WaitOne();
+        }
+
+        public void Stop()
+        {
+            try
+            {
+                _listenSocket.Close();
+            }
+            catch
+            {
+                // ignored
+            }
+
+            Mutex.ReleaseMutex();
+        }
+
+        private void OnIoCompleted(object sender, SocketAsyncEventArgs e)
+        {
+            switch (e.LastOperation)
+            {
+                case SocketAsyncOperation.Receive:
+                    ProcessReceive(e);
+                    break;
+                default:
+                    throw new ArgumentException("The last operation completed on the socket was not a receive");
+            }
+        }
+
+        private void StartAccept(SocketAsyncEventArgs acceptEventArg)
+        {
+            if (acceptEventArg == null)
+            {
+                acceptEventArg = new SocketAsyncEventArgs();
+                acceptEventArg.Completed += (sender, e) => ProcessAccept(e);
+            }
+            else
+            {
+                acceptEventArg.AcceptSocket = null;
+            }
+
+            _acceptedClientsSemaphore.WaitOne();
+            if (!_listenSocket.AcceptAsync(acceptEventArg)) ProcessAccept(acceptEventArg);
```

Comment:
This is throwing an unhandled exception when the application attempts to close itself by calling `Exit()` on the BlishHud.cs instance.

![image](https://user-images.githubusercontent.com/1950594/60750454-b8638d00-9f76-11e9-9217-e3bcd2a56aba.png)

Reply:
You can replicate this by right-clicking on the Blish HUD icon in your tray and selecting "Close Blish HUD."

Reply:
This is a known bug, since I first implemented it.
The GC is faster with cleaning up than my `_server.Stop()` and I have no solution currently.
I just have to find out where the try except has to go, because it doesnt crash at the same line all the time. It just crashes wherever it currently is.
Is on the TODO